core/src/main/java/jenkins/install/InstallUtil.java

```diff
@@ -0,0 +1,230 @@
+            File configFile = getConfigFile();
+            if (configFile.exists()) {
+                try {
+                    String lastVersion = XMLUtils.getValue("/hudson/version", configFile);
+                    if (lastVersion.length() > 0) {
+                        return lastVersion;
+                    }
+                } catch (Exception e) {
+                    LOGGER.log(SEVERE, "Unexpected error reading global config.xml", e);
+                }
+            }
+            return NEW_INSTALL_VERSION.toString();
+        }
+    }
+
+    /**
+     * Save a specific version as the last execute version.
+     * @param version The version to save.
+     */
+    static void saveLastExecVersion(@Nonnull String version) {
+        File lastExecVersionFile = getLastExecVersionFile();
+        try {
+            FileUtils.write(lastExecVersionFile, version);
+        } catch (IOException e) {
+            LOGGER.log(SEVERE, "Failed to save " + lastExecVersionFile.getAbsolutePath(), e);
+        }
+    }
+
+    static File getConfigFile() {
+        return new File(Jenkins.getActiveInstance().getRootDir(), "config.xml");
+    }
+
+    static File getLastExecVersionFile() {
+        return new File(Jenkins.getActiveInstance().getRootDir(), ".last_exec_version");
+    }
+
+    static File getInstallingPluginsFile() {
+        return new File(Jenkins.getActiveInstance().getRootDir(), ".installing_plugins");
+    }
+
+    private static String getCurrentExecVersion() {
+        if (Jenkins.VERSION.equals(Jenkins.UNCOMPUTED_VERSION)) {
+            // This should never happen!! Only adding this check in case someone moves the call to this method to the wrong place.
+            throw new IllegalStateException("Unexpected call to InstallUtil.getCurrentExecVersion(). Jenkins.VERSION has not been initialized. Call computeVersion() first.");
+        }
+        return Jenkins.VERSION;
+    }
+
+    /**
+     * Returns a list of any plugins that are persisted in the installing list
+     */
+    @SuppressWarnings("unchecked")
+	public static synchronized @CheckForNull Map<String,String> getPersistedInstallStatus() {
+        File installingPluginsFile = getInstallingPluginsFile();
+        if(installingPluginsFile == null || !installingPluginsFile.exists()) {
+		return null;
+        }
+        return (Map<String,String>)new XStream().fromXML(installingPluginsFile);
+    }
+
+    /**
+     * Persists a list of installing plugins; this is used in the case Jenkins fails mid-installation and needs to be restarted
+     * @param installingPlugins
+     */
+    public static synchronized void persistInstallStatus(List<UpdateCenterJob> installingPlugins) {
+        File installingPluginsFile = getInstallingPluginsFile();
+	if(installingPlugins == null || installingPlugins.isEmpty()) {
+		installingPluginsFile.delete();
+		return;
+	}
+	LOGGER.fine("Writing install state to: " + installingPluginsFile.getAbsolutePath());
```

Comment:
don't use string concatenation in logging messages wher the level is < `Level.INFO` - use parametrized logging - to avoid the unnecessary operation.